hicommonwealth/commonwealth
# Apply gating for actions

## Link to Issue
Closes: #5352

## Description of Changes
- Adds migration to increase `reject_reason` length to 1024
- Updates `updateTopic` route to allow updating group_ids
- Adds `FEATURE_FLAG_GROUP_CHECK_ENABLED` env var to switch validation implementation
- Applies group requirements check for
  - Thread creation
  - Comment creation
  - Thread reaction creation

## Test Plan

CA test locally:

Must test with 2 users. Assumes that your wallets do NOT contain USDC.

- Run migration
- Add env `FEATURE_FLAG_GROUP_CHECK_ENABLED=true`
- Set user as super admin
- Create new topic in Ethereum community
- Create new group: `POST http://localhost:8080/api/groups`
```json
{
    "author_chain": "ethereum",
    "chain": "ethereum",
    "address": "<ADDRESS>",
    "jwt": "<JWT>",
    "metadata": {
        "name": "Demo Group",
        "description": "Group"
    },
    "requirements": [
        {
            "rule": "threshold",
            "data": {
                "threshold": "10",
                "source": {
                    "source_type": "erc20",
                    "evm_chain_id": 1,
                    "contract_address": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48"
                }
            }
        }
    ],
    "topics": [<TOPIC_ID>]
}
```
- Unset user as super admin and ensure user is not community mod/admin
- Create thread in community– confirm that this FAILS due to insufficient balance
- Set user as super admin and create the thread– confirm this succeeds
- Switch to another user (not admin)
- Go to the thread and create a comment– confirm that this FAILS due to insufficient balance
- Create a reaction on the thread– confirm that this FAILS due to insufficient balance

NOTE: It's important to switch users because if the user is the owner of the thread, it will bypass the membership check.

## Deployment Plan
- Run migration
- Add env `FEATURE_FLAG_GROUP_CHECK_ENABLED=true`

## Other Considerations
- Check does not apply for comment reaction creation.
